src/log/main.cr

```diff
@@ -0,0 +1,142 @@
+
+    # remove generic arguments
+    paren = source.index('(')
+    source = source[0...paren] if paren
+
+    ::Log.for(source, level)
+  end
+
+  # :ditto:
+  def for(type : Class, level : Severity? = nil) : Log
+    ::Log.for(type, level)
+  end
+
+  private Top = Log.for("")
+
+  {% for method in %i(debug verbose info warn error fatal) %}
+    # See `Log#{{method.id}}`.
+    def self.{{method.id}}(*, exception : Exception? = nil)
+      Top.{{method.id}}(exception: exception) do
+        yield
+      end
+    end
+  {% end %}
+
+  @@builder = Builder.new
+
+  # Returns the default `Log::Builder` used for `Log.for` calls.
+  def self.builder
+    @@builder
+  end
+
+  # Returns the current fiber logging context.
+  def self.context : Log::Context
+    Fiber.current.logging_context
+  end
+
+  # Sets the current fiber logging context.
+  def self.context=(value : Log::Context)
+    Fiber.current.logging_context = value
+  end
+
+  # Returns the current fiber logging context.
+  def context : Log::Context
+    Log.context
+  end
+
+  # Sets the current fiber logging context.
+  def context=(value : Log::Context)
+    Log.context = value
+  end
+
+  class Context
+    # Clears the current `Fiber` logging context.
+    #
+    # ```
+    # Log.context.clear
+    # Log.info { "message with empty context" }
+    # ```
+    def clear
+      Fiber.current.logging_context = Log::Context.new
+    end
+
+    # Extends the current `Fiber` logging context.
+    #
+    # ```
+    # Log.context.set a: 1
+    # Log.context.set b: 2
+    # Log.info { %q(message with {"a" => 1, "b" => 2} context") }
+    # extra = {:c => "3"}
+    # Log.context.set extra
+    # Log.info { %q(message with {"a" => 1, "b" => 2, "c" => "3" } context) }
+    # extra = {"c" => 3}
+    # Log.context.set extra
+    # Log.info { %q(message with {"a" => 1, "b" => 2, "c" => 3 } context) }
+    # ```
+    def set(**kwargs)
+      extend_fiber_context(Fiber.current, Log::Context.new(kwargs))
+    end
+
+    # :ditto:
+    def set(values : Hash(String, V)) forall V
+      extend_fiber_context(Fiber.current, Log::Context.new(values))
+    end
+
+    # :ditto:
+    def set(values : Hash(Symbol, V)) forall V
+      extend_fiber_context(Fiber.current, Log::Context.new(values))
+    end
+
+    # :ditto:
+    def set(values : NamedTuple) forall V
+      extend_fiber_context(Fiber.current, Log::Context.new(values))
+    end
+
+    private def extend_fiber_context(fiber : Fiber, values : Context) forall V
+      context = fiber.logging_context
+      fiber.logging_context = context.merge(values)
+    end
+
+    # Method to save and restore the current logging context.
+    #
+    # ```
+    # Log.context.set a: 1
+    # Log.info { "message with {"a" => 1} context" }
+    # Log.using do
+    #   Log.context.set b: 2
+    #   Log.info { "message with {"a" => 1, "b" => 2} context" }
+    # end
+    # Log.info { "message with {"a" => 1} context" }
+    # ```
+    def using
```

Comment:
"using" is very unspecific. I'd call it `with_context`.

Reply:
But it is used as `Log.context.using do`
I am used that using handles resources or disposable since .NET.

`Log.context.with` could be an alternative, but it reads more weird to me. And is using a keyword as a method name. I prefer to avoid if possible.

Meanwhile I notice the sample had some errors due to symbol to string refactor... fixing those.

Reply:
What about `Log.with_context do`?

Reply:
I can live with that :-)

Reply:
done